File: src/hubs/hub7_dental_operations/xray_caries_sentinel.py
"""X-Ray Caries Sentinel (Agent 27, Hub 7)."""

import hashlib
import os
import re
from datetime import datetime, timezone

SUBJECT_ID_PEPPER = os.environ.get("SUBJECT_ID_PEPPER", "CHANGE_ME_IN_PRODUCTION")

XRAY_ALLOWLIST = ["tooth_number", "caries_confidence_score"]

FIELD_PATTERNS = {
    "patient_name": re.compile(r"patient name:\s*([^\n]+)", re.IGNORECASE),
    "patient_id": re.compile(r"patient id:\s*([^\n]+)", re.IGNORECASE),
    "tooth_number": re.compile(r"tooth number:\s*([^\n]+)", re.IGNORECASE),
    "caries_confidence_score": re.compile(r"caries confidence score:\s*(\d*\.?\d+)", re.IGNORECASE),
}

HIGH_DECAY_THRESHOLD = 0.85


def hash_subject_id(patient_name, patient_id):
    return hashlib.sha256(f"{SUBJECT_ID_PEPPER}:{patient_name}:{patient_id}".encode()).hexdigest()


def mask_value(value):
    s = str(value) if value is not None else ""
    if len(s) <= 2:
        return "*" * len(s)
    return f"{s[0]}{'*' * (len(s) - 2)}{s[-1]}"


def iso_now():
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def extract_field(text, pattern):
    match = pattern.search(text)
    return match.group(1).strip() if match else ""


class XrayCariesSentinel:
    # Stage 1 (Bitewing Annotation Ingestion): patient name/ID are only
    # used to derive a hashed subject_id; only tooth number and the CV
    # classifier's confidence score survive into the record.
    def ingest(self, raw_text):
        if not isinstance(raw_text, str) or not raw_text.strip():
            raise TypeError("raw_text must be a non-empty string")
        patient_name = extract_field(raw_text, FIELD_PATTERNS["patient_name"])
        patient_id = extract_field(raw_text, FIELD_PATTERNS["patient_id"])
        if not patient_name or not patient_id:
            raise ValueError("raw_text requires Patient Name and Patient ID to derive a subject_id")
        tooth_number = extract_field(raw_text, FIELD_PATTERNS["tooth_number"])
        if not tooth_number:
            raise ValueError("raw_text requires Tooth Number")

        record = {
            "subject_id": hash_subject_id(patient_name, patient_id),
            "tooth_number": tooth_number,
            "caries_confidence_score": float(extract_field(raw_text, FIELD_PATTERNS["caries_confidence_score"]) or 0),
        }

        for key in list(record.keys()):
            if key != "subject_id" and key not in XRAY_ALLOWLIST:
                del record[key]
        return record

    # Stage 2 (Compliance & Decay Risk Screening): defense-in-depth guard,
    # then flag a high-confidence decay finding.
    def evaluate_decay_risk(self, record):
        leaked = next((k for k in record if k != "subject_id" and k not in XRAY_ALLOWLIST), None)
        if leaked:
            raise ValueError(f'Compliance violation: unexpected field "{leaked}" reached decay risk screening stage')

        high_decay = record["caries_confidence_score"] >= HIGH_DECAY_THRESHOLD
        return {"high_decay": high_decay, "status": "HIGH_DECAY_DETECTION" if high_decay else "ROUTINE_FINDING"}

    # Stage 3 (Integration): only reachable on a high-confidence finding —
    # an interactive, patient-facing case presentation to support informed
    # treatment consent.
    def build_patient_case_visualization(self, record, decay_result):
        return (
            f'CASE-PRESENTATION|{record["subject_id"]}|TOOTH:{record["tooth_number"]}|FINDING:{decay_result["status"]}|'
            f'CONFIDENCE:{record["caries_confidence_score"]}|VISUAL_AID:INTERACTIVE_3D_TOOTH_OVERLAY|'
            f'PATIENT_EXPLANATION:This X-ray shows a high-confidence area of decay on tooth {record["tooth_number"]} '
            f"that our AI imaging analysis flagged for review with your dentist.|GENERATED_AT:{iso_now()}"
        )

    # Stage 4 (Audit Telemetry): masked Splunk SIEM event on every reading.
    def build_splunk_event(self, record, decay_result, outcome):
        return {
            "time": datetime.now(timezone.utc).timestamp(),
            "host": "xray-caries-sentinel",
            "source": "hub7_dental_operations/xray_caries_sentinel",
            "sourcetype": "_json",
            "event": {
                "subject_id": record["subject_id"],
                "action": outcome,
                "status": decay_result["status"],
                "caries_confidence_score": record["caries_confidence_score"],
                "masked_tooth_number": mask_value(record["tooth_number"]),
                "processed_at": iso_now(),
            },
        }

    def run(self, raw_text):
        record = self.ingest(raw_text)
        decay_result = self.evaluate_decay_risk(record)
        case_presentation = self.build_patient_case_visualization(record, decay_result) if decay_result["high_decay"] else None
        splunk_event = self.build_splunk_event(record, decay_result, "high_decay_case_presented" if case_presentation else "routine_finding_logged")
        return {"case_presentation": case_presentation, "status": decay_result["status"], "splunk_event": splunk_event}

File: src/hubs/hub7_dental_operations/xray_caries_sentinel.js
'use strict';

const crypto = require('crypto');

const SUBJECT_ID_PEPPER = process.env.SUBJECT_ID_PEPPER || 'CHANGE_ME_IN_PRODUCTION';

const XRAY_ALLOWLIST = ['toothNumber', 'cariesConfidenceScore'];

const FIELD_PATTERNS = {
  patientName: /patient name:\s*([^\n]+)/i,
  patientId: /patient id:\s*([^\n]+)/i,
  toothNumber: /tooth number:\s*([^\n]+)/i,
  cariesConfidenceScore: /caries confidence score:\s*(\d*\.?\d+)/i,
};

const HIGH_DECAY_THRESHOLD = 0.85;

function hashSubjectId(patientName, patientId) {
  return crypto
    .createHash('sha256')
    .update(`${SUBJECT_ID_PEPPER}:${patientName}:${patientId}`)
    .digest('hex');
}

function maskValue(value) {
  const str = String(value ?? '');
  if (str.length <= 2) return '*'.repeat(str.length);
  return `${str[0]}${'*'.repeat(str.length - 2)}${str[str.length - 1]}`;
}

function extractField(text, pattern) {
  const match = text.match(pattern);
  return match ? match[1].trim() : '';
}

class XrayCariesSentinel {
  // Stage 1 (Bitewing Annotation Ingestion): patient name/ID are only
  // used to derive a hashed subjectId; only tooth number and the CV
  // classifier's confidence score survive into the record.
  ingest(rawText) {
    if (typeof rawText !== 'string' || !rawText.trim()) {
      throw new TypeError('rawText must be a non-empty string');
    }
    const patientName = extractField(rawText, FIELD_PATTERNS.patientName);
    const patientId = extractField(rawText, FIELD_PATTERNS.patientId);
    if (!patientName || !patientId) {
      throw new Error('rawText requires Patient Name and Patient ID to derive a subjectId');
    }
    const toothNumber = extractField(rawText, FIELD_PATTERNS.toothNumber);
    if (!toothNumber) {
      throw new Error('rawText requires Tooth Number');
    }

    const record = {
      subjectId: hashSubjectId(patientName, patientId),
      toothNumber,
      cariesConfidenceScore: parseFloat(extractField(rawText, FIELD_PATTERNS.cariesConfidenceScore)) || 0,
    };

    for (const key of Object.keys(record)) {
      if (key !== 'subjectId' && !XRAY_ALLOWLIST.includes(key)) {
        delete record[key];
      }
    }
    return record;
  }

  // Stage 2 (Compliance & Decay Risk Screening): defense-in-depth guard,
  // then flag a high-confidence decay finding.
  evaluateDecayRisk(record) {
    const leaked = Object.keys(record).find((key) => key !== 'subjectId' && !XRAY_ALLOWLIST.includes(key));
    if (leaked) {
      throw new Error(`Compliance violation: unexpected field "${leaked}" reached decay risk screening stage`);
    }

    const highDecay = record.cariesConfidenceScore >= HIGH_DECAY_THRESHOLD;
    return { highDecay, status: highDecay ? 'HIGH_DECAY_DETECTION' : 'ROUTINE_FINDING' };
  }

  // Stage 3 (Integration): only reachable on a high-confidence finding —
  // an interactive, patient-facing case presentation to support informed
  // treatment consent.
  buildPatientCaseVisualization(record, decayResult) {
    return `CASE-PRESENTATION|${record.subjectId}|TOOTH:${record.toothNumber}|FINDING:${decayResult.status}|CONFIDENCE:${record.cariesConfidenceScore}|VISUAL_AID:INTERACTIVE_3D_TOOTH_OVERLAY|PATIENT_EXPLANATION:This X-ray shows a high-confidence area of decay on tooth ${record.toothNumber} that our AI imaging analysis flagged for review with your dentist.|GENERATED_AT:${new Date().toISOString()}`;
  }

  // Stage 4 (Audit Telemetry): masked Splunk SIEM event on every reading.
  buildSplunkEvent(record, decayResult, outcome) {
    return {
      time: Date.now() / 1000,
      host: 'xray-caries-sentinel',
      source: 'hub7_dental_operations/xray_caries_sentinel',
      sourcetype: '_json',
      event: {
        subjectId: record.subjectId,
        action: outcome,
        status: decayResult.status,
        cariesConfidenceScore: record.cariesConfidenceScore,
        maskedToothNumber: maskValue(record.toothNumber),
        processedAt: new Date().toISOString(),
      },
    };
  }

  run(rawText) {
    const record = this.ingest(rawText);
    const decayResult = this.evaluateDecayRisk(record);
    const casePresentation = decayResult.highDecay ? this.buildPatientCaseVisualization(record, decayResult) : null;
    const splunkEvent = this.buildSplunkEvent(record, decayResult, casePresentation ? 'high_decay_case_presented' : 'routine_finding_logged');
    return { casePresentation, status: decayResult.status, splunkEvent };
  }
}

module.exports = {
  XrayCariesSentinel,
  HIGH_DECAY_THRESHOLD,
  XRAY_ALLOWLIST,
};
